docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,108 @@
+Cucumber Core is the [inner hexagon](https://en.wikipedia.org/wiki/Hexagonal_architecture_(software)) for the [Ruby flavour of Cucumber](https://github.com/cucumber/cucumber-ruby). It contains the core domain logic to execute Cucumber features. It has no user interface, just a Ruby API.
+
+## Overview
+
+The entry-point is a single method on the module `Cucumber::Core` called
+`#execute`. Here's what it does:
+
+1. Parses the plain-text Gherkin documents into an **AST**
+2. Compiles the AST down to **test cases**
+3. Passes the test cases through any **filters**
+4. Executes the test cases, emitting **events** as it goes
+
+### The AST
+
+The Abstract Syntax Tree or AST is an object graph that represents the Gherkin documents you've passed into the core. Things like `Feature`, `Scenario` and `ExamplesTable`.
+
+These are immutable value objects.
+
+### Test cases
+
+Your Gherkin might contain scenarios, as well as examples from tables beneath a scenario outline.
+
+Test cases represent the general case of both of these. We compile the AST down to instances of `Cucumber::Core::Test::Case`, each containing a number of instances of `Cucumber::Core::Test::Step`. It's these that are then filtered and executed.
+
+Test cases and their test steps are immutable value objects.
+
+### Filters
+
+Once we have the test cases, and they've been activated by the mappings, you may want to pass them through a filter or two. Filters can be used to do things like activate, sort, replace or remove some of the test cases or their steps before they're executed.
+
+### Events
+
+Events are how you find out what is happening during your test run. As the test cases and steps are executed, the runner emits events to signal what's going on.
+
+The following events are emitted during a run:
```

Comment:
The following, is a non-exhaustive list of the events that are emitted during a run

OR

Some of the events that can be emitted during a run are: